RemotePlayerManager port (from Monumenta Network Chat)

Ascynx opened the pull request:
On top of the port:
	 - Added the ability to register plugin data on the remote player messages.
	 - Added event GatherRemotePlayerDataEvent to gather the plugin data to be added on the remote player messages.
	 - Added events RemotePlayerLoadedEvent & RemotePlayerUnloadedEvent to get an easier way to listen to changes in the full online player list.

Ascynx committed:
RemotePlayerManager Port

Added the ability to register plugin data on the remote player messages.

Added event GatherRemotePlayerDataEvent to gather the plugin data to be added on the remote player messages.

Added events RemotePlayerLoadedEvent & RemotePlayerUnloadedEvent to get an easier way to listen to changes in the full online player list.

NickNackGus commented:
Just marking myself a reviewer; I still need to get to this later.

Combustible commented on src/main/java/com/playmonumenta/networkrelay/RemotePlayerManager.java:
```diff
@@ -0,0 +1,347 @@
+
+
+	private final NetworkRelay mPlugin;
+	protected RemotePlayerManager(NetworkRelay plugin) {
+        mPlugin = plugin;
+		String lShard = getShardName();
+		try {
+			for (String shard: NetworkRelayAPI.getOnlineShardNames()) {
+				if (shard.equals(lShard)) {
+					continue;
+				}
+				mRemotePlayerShardMapped.put(shard, new ConcurrentSkipListMap<>());
+			}
+		} catch (Exception ex) {
+			mPlugin.getLogger().warning("Failed to get load shards");
+		}
+
+		try {
+			NetworkRelayAPI.sendBroadcastMessage(REMOTE_PLAYER_REFRESH_CHANNEL, new JsonObject());
+		} catch (Exception e) {
+			mPlugin.getLogger().warning("Failed to broadcast to channel " + REMOTE_PLAYER_REFRESH_CHANNEL);
+        }
+
+		INSTANCE = this;
+    }
```
Weird indentation - maybe tabs/spaces issues?

Ascynx replied:
don't believe so, no idea where it's coming from, but might be fixed in next commit

Combustible commented on src/main/java/com/playmonumenta/networkrelay/RemotePlayerManager.java:
```diff
@@ -0,0 +1,347 @@
+
+		@Nullable
+		public JsonObject getPluginData(String pluginId) {
+			if (mPluginData.containsKey(pluginId)) {
+				return mPluginData.get(pluginId);
+			}
+			return null;
```
should just return the get() directly, since it will return null if it doesn't exist.

Combustible commented on src/main/java/com/playmonumenta/networkrelay/GatherRemotePlayerDataEvent.java:
```diff
@@ -0,0 +1,59 @@
+import java.util.Map;
+import org.bukkit.event.Event;
+import org.bukkit.event.HandlerList;
+import org.jetbrains.annotations.NotNull;
```
Can you remove all the instances of NotNull? This is the default, if it's not @ Nullable then it's automatically assumed to be NotNull.

Ascynx replied:
Was automatically added by intelliJ
![image](https://github.com/TeamMonumenta/monumenta-network-relay/assets/78341107/cc7a6261-4f35-4e83-8e94-8bc65a6519d8)

Combustible commented on src/main/java/com/playmonumenta/networkrelay/RemotePlayerManager.java:
```diff
@@ -0,0 +1,347 @@
+		return INSTANCE;
+	}
+
+	protected String[] getAllOnlinePlayersName() {
+		return mRemotePlayersByName.keySet().toArray(new String[0]);
+	}
```
Is there some reason we're converting this to an array here? Why not just return the keySet() ?

Ascynx replied:
Command suggestions

Ascynx replied:
but it's not like it can't get transformed into an array from the other plugin's side.

NickNackGus replied:
I've also made a habit of not returning original collections in order to ensure other code cannot modify internal data structures in unexpected ways. I believe this was originally type `List<String>` instead, but the amount of code changes between this and the original makes it painful to review.

Ascynx replied:
![image](https://github.com/TeamMonumenta/monumenta-network-relay/assets/78341107/7991e8e8-b8bd-4e34-aed8-965e15132618)
it was not

Combustible commented on src/main/java/com/playmonumenta/networkrelay/RemotePlayerManager.java:
```diff
@@ -0,0 +1,347 @@
+	}
+
+	@Nullable
+	protected RemotePlayer getRemotePlayer(@Nullable UUID playerUuid) {
+		if (playerUuid == null) {
+			return null;
+		}
+
+		@Nullable RemotePlayer remotePlayer = mRemotePlayersByUuid.get(playerUuid);
+		return remotePlayer;
+	}
```
This whole block is weird - seems like playerUuid should be not nullable (i.e. no annotation), and just return directly `mRemovePlayersByUuid.get(playerUuid)` without the temporary variable.

Combustible commented on src/main/java/com/playmonumenta/networkrelay/RemotePlayerManager.java:
```diff
@@ -0,0 +1,347 @@
+
+			RemotePlayerLoadedEvent remotePLE = new RemotePlayerLoadedEvent(remotePlayer);
+			Bukkit.getServer().getPluginManager().callEvent(remotePLE);
+		} else if (!Objects.equals(getShardName(), remotePlayer.mName)) {
+			mPlugin.getLogger().fine("Detected race condition, triggering refresh on " + remotePlayer.mName);
+			@Nullable Player localPlayer = Bukkit.getPlayer(remotePlayer.mUuid);
+			if (localPlayer != null) {
+				refreshLocalPlayer(localPlayer);
+			}
+		}
```
This logic doesn't make sense to me - why are we comparing the shard name and the player name?

Ascynx replied:
doesn't make sense to me, but if you don't have it, when someone changes shard, the remote update arrives before the local quit event (resulting in the entire entry being deleted everywhere)

NickNackGus replied:
Correct - transferring shards is a sequence of:
- Save the player on src shard
- Load the player on dst shard and log in
- The player's control switches shards
- Log out the player on src shard and log out

So there needs to be a check for if the player is actually on the shard they're logging out from.

Ascynx replied:
I just noticed that I was checking the player name,
guessing I just used tab autocomplete and forgot to check if it was correct.

Ascynx committed:
PR Reviews part 1

Added back vanished default variables

Switched to using Optionals for values where the manager could be null.

A shard switch will now only be 1 load/unload event, even when a race condition would happen.

Ascynx committed:
Moved stuff around to make it follow how it was formatted in the network-chat counterpart.

Ascynx committed:
Switched to fully static (and so no more optional)

Ascynx committed:
Forgot about 2-3 things

NickNackGus committed:
Code review changes

NickNackGus commented on src/main/java/com/playmonumenta/networkrelay/RemotePlayerManager.java:
```diff
@@ -33,10 +36,17 @@ public static class RemotePlayer {
+			mShard = getShardName();
 			mPluginData = new ConcurrentHashMap<>();
+
+			// Gather Plugin data
```
Moved plugin data collection here. It should only update when updated by the refresh method.

NickNackGus commented on src/main/java/com/playmonumenta/networkrelay/RemotePlayerManager.java:
```diff
@@ -45,25 +55,27 @@ public RemotePlayer(JsonObject remoteData) {
 		}
 
-		private ConcurrentMap<String, JsonObject> serializePluginData(JsonObject remoteData) {
+		private ConcurrentMap<String, JsonObject> deserializePluginData(JsonObject remoteData) {
```
Serialize -> Convert to string (or json in this case)
Deserialize -> Convert from string

Ascynx replied:
I knew I would make that mistake at least once in my life

NickNackGus commented on src/main/java/com/playmonumenta/networkrelay/RemotePlayerManager.java:
```diff
@@ -78,26 +90,19 @@ public void broadcast() {
 		}
 	}
 
+	public static final int REMOTE_PLAYER_MESSAGE_TTL = 5;
```
We don't want to keep these messages around for very long - if they're too far out of date, they should be discarded sooner. This should probably move to a config, but here works for now.

NickNackGus commented on src/main/java/com/playmonumenta/networkrelay/RemotePlayerManager.java:
```diff
@@ -108,100 +113,105 @@ public void broadcast() {
 	private static final Set<UUID> mVisiblePlayers = new ConcurrentSkipListSet<>();
 
-	protected RemotePlayerManager() {
+	private RemotePlayerManager() {
```
We want to prevent more than one instance from being created, even in the plugin itself. Switched back to `getInstance()` style.

U5B replied:
Would say to move those constants up towards the top of the file, but probably isn't necessary

NickNackGus commented on src/main/java/com/playmonumenta/networkrelay/RemotePlayerManager.java:
```diff
@@ -108,100 +113,105 @@ public void broadcast() {
 		} catch (Exception ex) {
-			CustomLogger.getInstance().ifPresent((logger) -> logger.warning("Failed to get load shards"));
+			MMLog.severe("Failed to get remote shard names");
+			throw new RuntimeException("Failed to get remote shard names");
```
Hitting this point requires a critical error to have occurred.

Ascynx committed:
nullables go brrrr

NickNackGus commented on src/main/java/com/playmonumenta/networkrelay/RemotePlayerManager.java:
```diff
@@ -108,100 +113,105 @@ public void broadcast() {
 
-	@Nullable
-	protected static RemotePlayerManager getInstance() {
+	public static RemotePlayerManager getInstance() {
```
This and many other methods were originally public so that they could be used by other plugins. This is more important now that we're trying to use this as an API.

Ascynx replied:
I made them protected so people used the methods in NetworkRelayAPI instead of the class itself.

NickNackGus replied:
Oh, I missed those. I can mark them protected again then.

NickNackGus commented on src/main/java/com/playmonumenta/networkrelay/RemotePlayerManager.java:
```diff
@@ -108,100 +113,105 @@ public void broadcast() {
+			MMLog.severe("Failed to get shard name");
+		}
+		if (shardName == null) {
+			throw new RuntimeException("Got null shard name");
```
A critical error would have to happen to reach this point as well. There is no valid case where we return null.

NickNackGus commented on src/main/java/com/playmonumenta/networkrelay/RemotePlayerManager.java:
```diff
@@ -108,100 +113,105 @@ public void broadcast() {
 	}
 
-	protected static Set<String> getAllOnlinePlayersName(boolean visibleOnly) {
+	public static Set<String> getAllOnlinePlayersName(boolean visibleOnly) {
```
This was originally for all online players, regardless of visibility, with a separate method for only visible players that was kept below and used internally. I'd keep this as it was before, but I figure requiring a visibility parameter makes it more clear for people using the API.

NickNackGus commented on src/main/java/com/playmonumenta/networkrelay/RemotePlayerManager.java:
```diff
@@ -108,100 +113,105 @@ public void broadcast() {
 			return getVisiblePlayerNames();
 		}
-		return mRemotePlayersByName.keySet();
+		return new HashSet<>(mRemotePlayersByName.keySet());
```
Avoid exposing modify access to external plugins - we don't want them breaking our internal state. (There may be a better way to do this, not sure what it would be.)

NickNackGus commented on src/main/java/com/playmonumenta/networkrelay/RemotePlayerManager.java:
```diff
@@ -278,12 +289,15 @@ protected static void refreshLocalPlayer(Player player) {
 		}
 		remotePlayer.broadcast();
+
+		RemotePlayerLoadedEvent remotePLE = new RemotePlayerLoadedEvent(remotePlayer);
```
If I understand right, this event is supposed to fire every time player state changes and the player is loaded, even if they were already online - at least this is how it worked for remote players. We can adjust it if needed.

NickNackGus commented on src/main/java/com/playmonumenta/networkrelay/RemotePlayerManager.java:
```diff
@@ -278,12 +289,15 @@ protected static void refreshLocalPlayer(Player player) {
 		@Nullable RemotePlayer lastPlayerState = mRemotePlayersByUuid.get(playerUuid);
 		if (lastPlayerState != null) {
-			CustomLogger.getInstance().ifPresent(logger -> logger.fine("Unregistering local player"));
+			MMLog.fine("Unregistering player " + lastPlayerState.mName);
```
Player may not be local (on this shard).

NickNackGus commented on src/main/java/com/playmonumenta/networkrelay/RemotePlayerManager.java:
```diff
@@ -313,21 +329,33 @@ private static void remotePlayerChange(JsonObject data) {
 			mRemotePlayersByName.put(remotePlayer.mName, remotePlayer);
+			if (!remotePlayer.mIsHidden) {
+				mVisiblePlayers.add(remotePlayer.mUuid);
+			}
```
Ensure hidden state is tracked for remote players, not just local players.

NickNackGus commented on src/main/java/com/playmonumenta/networkrelay/RemotePlayerManager.java:
```diff
@@ -313,21 +329,33 @@ private static void remotePlayerChange(JsonObject data) {
+			// Player logged off on remote shard, but is locally online.
+			// This can happen if the remote shard was not notified the player logged in here in time.
+			MMLog.fine("Detected race condition, triggering refresh on " + remotePlayer.mName);
+			refreshLocalPlayer(localPlayer);
```
I did see the special case for when a local player quits, but is online on another shard. This is a fail safe for if the quit event arrives before that shard knows the player logged in elsewhere, which is what made this a race condition.

NickNackGus commented on src/main/java/com/playmonumenta/networkrelay/RemotePlayerManager.java:
```diff
@@ -339,33 +367,39 @@ public void destOfflineEvent(DestOfflineEvent event) {
+			|| command.contains("vanish")) {
+			Bukkit.getScheduler().runTask(NetworkRelay.getInstance(), RemotePlayerManager::refreshLocalPlayers);
+		}
+	}
```
I agree the `/team` check isn't required here, since we aren't using the player name chat component. We do want to keep the vanish state in sync, however (which I know wasn't in the first attempt to port this).

NickNackGus commented on src/main/java/com/playmonumenta/networkrelay/RemotePlayerManager.java:
```diff
@@ -339,33 +367,39 @@ public void destOfflineEvent(DestOfflineEvent event) {
-		if (remotePlayer != null) {
-			RemotePlayerLoadedEvent remotePLE = new RemotePlayerLoadedEvent(remotePlayer);
-			Bukkit.getServer().getPluginManager().callEvent(remotePLE);
-		}
```
Moved to the `refreshLocalPlayer` method.

NickNackGus commented on src/main/java/com/playmonumenta/networkrelay/util/MMLog.java:
```diff
@@ -0,0 +1,70 @@
+import java.util.function.Supplier;
+import java.util.logging.Level;
+
+public class MMLog {
```
I'm surprised we didn't add this to NetworkRelay years ago - it saves you from needing to get the custom logger instance all the time and check if it's valid. I'm also confused as to why `CustomLogger.getInstance()` returns an `Optional` instead of a `@Nullable`. None of that is your fault, and I see why it would make copying it frustrating.

Ascynx replied:
The Optional part was a me thing, there wasn't an getInstance method nor INSTANCE field before.
I just don't like null checks.

NickNackGus committed:
Fixed missing player name in log message

NickNackGus committed:
Fixed unused imports

NickNackGus committed:
Merge branch 'ascynx/rpm-port' of github.com:Ascynx/monumenta-network-relay into ascynx/rpm-port

NickNackGus commented:
I'd still like to figure out how to make it so these values can be written by Bungee/Paper and read by Bungee/Paper/Generic, rather than being Paper-only. I suspect that'll involve the abstraction style classes that are already present in this plugin.

NickNackGus committed:
Marked many API-accessible methods as protected again

Ascynx committed:
Abstracted RemotePlayerManager

(surely I did it correctly)

Ascynx committed:
override

Ascynx committed:
forgot about nullables

Ascynx committed:
reviewdoggo did not like it

Ascynx committed:
maybe this will do

Ascynx committed:
tab indents this time around

Ascynx committed:
Moved remote player attributes around

Moved vanished to RemotePlayerPaper

Moved Plugin data to RemotePlayerAbstraction

Added Proxy name to RemotePlayerBungee (should be moved to the Bungee implementation of RemotePlayerManager once made)

Ascynx committed:
oops forgot about that

Ascynx committed:
does it want the abstract before static?

NickNackGus committed:
Updated RemotePlayer data types to be loadable from anywhere; RemotePlayerPaper can be created locally on Paper

NickNackGus commented:
I've pushed a commit to make it possible to read RemotePlayer<Whatever> from any server type as intended. The tracking of remote player data needs to be moved out of the Paper class, since all shards will be tracking that in the same way. The RemotePlayerManagerPaper class still needs to listen for Paper events and send updates for local players to be processed by everywhere else remotely, just like the Bungee equivalent will do for information from Bungee. Each RemotePlayer<Whatever> needs to be tracked in a map sorted by ServerType, which is included in each RemotePlayer<Whatever>.

Ascynx commented:
yk that makes me think it would probably be useful to allow reviewdog to review prs as we wouldn't have to look through its logs

Ascynx commented:
![image](https://github.com/TeamMonumenta/monumenta-network-relay/assets/78341107/fb6a02be-418b-40e6-aab3-fe6cbbe654e3)
ye ok this time it just didn't compile

NickNackGus committed:
Fixed missing argument

U5B committed:
usb: expose a way to get all player uuids

Signed-off-by: U5B <56985400+U5B@users.noreply.github.com>

Ascynx committed:
Merge branch 'TeamMonumenta:master' into ascynx/rpm-port

U5B commented:
Closed in favor of https://github.com/TeamMonumenta/monumenta-network-relay/pull/23
